power_grid_model_c/power_grid_model/include/power_grid_model/job_interface.hpp

```diff
@@ -14,75 +14,71 @@
 #include <utility>
 
 namespace power_grid_model {
-template <typename Adapter> class JobInterface {
+class JobInterface {
   public:
     // the multiple  NOSONARs are used to avoid the complaints about the unnamed concepts
-    template <typename ResultDataset>
-    void calculate(ResultDataset const& result_data, Idx pos = 0)
-        requires requires(Adapter& adapter) { // NOSONAR
-            { adapter.calculate_impl(result_data, pos) } -> std::same_as<void>;
+    template <typename Self, typename ResultDataset>
+    void calculate(this Self const& self, ResultDataset const& result_data, Idx pos = 0)
+        requires requires(this Self const& self) { // NOSONAR
+            { self.calculate_impl(result_data, pos) } -> std::same_as<void>;
         }
     {
-        return static_cast<Adapter*>(this)->calculate_impl(result_data, pos);
+        return self.calculate_impl(result_data, pos);
     }
 
-    void cache_calculate()
-        requires requires(Adapter& adapter) { // NOSONAR
-            { adapter.cache_calculate_impl() } -> std::same_as<void>;
+    template <typename Self>
+    void cache_calculate(this Self const& self)
+        requires requires(this Self const& self) { // NOSONAR
```

Comment:
We have build fails though because "explicit object parameter in requires" is not allowed. Need to think a bit more on how to tackle this. Do you know @mgovers ?

Reply:
Oh I see it. That extra `this` in `requires` is redundant. as the function parameter does the deduction.